lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -399,44 +430,14 @@ class swift::ParseableInterfaceBuilder {
         if (!buf) return true;
         uint64_t hash = xxHash64(buf->getBuffer());
         Deps.push_back(
-          FileDependency::hashBased(DepName, Status->getSize(), hash));
+          FileDependency::hashBased(DepNameToStore, IsSDKRelative,
+                                    Status->getSize(), hash));
       } else {
         uint64_t mtime =
           Status->getLastModificationTime().time_since_epoch().count();
         Deps.push_back(
-          FileDependency::modTimeBased(DepName, Status->getSize(), mtime));
-      }
-
-      if (moduleCachePath.empty())
-        continue;
-
-      // If Dep is itself a .swiftmodule in the cache dir, pull out its deps
-      // and include them in our own, so we have a single-file view of
-      // transitive deps: removes redundancies, and avoids opening and reading
-      // multiple swiftmodules during future loads.
```

Comment:
It seems like pulling these in was redundant? I think everything still works because whenever we load/depend on something from the cache we end up adding its serialized dependencies to the dependency tracker when validating that it's up to date, so they show up as one of the initial dependencies here and get serialized.

Reply:
That doesn't sound right. What if it's out of date and then we rebuild it and discover it doesn't depend on some of those files anymore?

Reply:
`ParseableInterfaceModuleLoaderImpl::dependenciesAreUpToDate()` is called by both `forwardingModuleIsUpToDate()` and `swiftModuleIsUpDate()` and adds the dependencies to the tracker regardless of whether they are out of date or not, so looks like that's a bug. I'll add a test for it and update the code to only add them to the tracker if they're all up to date.

In the case that they're not up to date though, we then go down the path of generating the swiftmodule from the interface again, and that adds all the `SubInstance`'s dependencies to the tracker (in `collectDepsForSerialization()`). Since all of *its* dependencies must be either up to date or rebuilt too, I think us pulling out the serialized deps here is redundant regardless. Does that make sense, or am I still missing something?

Reply:
I'm surprised that `dependenciesAreUpToDate` adds dependencies to the outer tracker at all. That doesn't seem like its job, and that would explain why this code is here. But having it do that does save us a bit of effort in reading from the cached compiled module twice. It at least deserves to be commented loudly, though, since that name sounds like a predicate.

Reply:
Yeah, `dependenciesAreUpToDate` should probably be renamed here. Good catch noticing that it doesn't check if they actually are up to date 😳

Reply:
I ended up making it not add the dependencies and added them a few levels up in `findOrBuildLoadableModule` and updated it's doc to mention that it adds the dependencies to the dependencyTracker too.